File: internal/worktree/slug.go
package worktree

import (
	"fmt"
	"regexp"
	"strings"
)

var (
	slugRegex     = regexp.MustCompile(`[^a-zA-Z0-9_-]+`)
	maxSlugLength = 80
)

// GenerateSlug creates filesystem-safe slug from branch name
func GenerateSlug(branch string) string {
	// Replace / with _
	slug := strings.ReplaceAll(branch, "/", "_")

	// Remove unsafe characters
	slug = slugRegex.ReplaceAllString(slug, "_")

	// Collapse multiple underscores
	slug = regexp.MustCompile(`_+`).ReplaceAllString(slug, "_")

	// Clean up underscores adjacent to dashes
	slug = regexp.MustCompile(`_-|-_`).ReplaceAllString(slug, "-")

	// Trim leading/trailing underscores
	slug = strings.Trim(slug, "_")

	// Truncate if too long
	if len(slug) > maxSlugLength {
		slug = slug[:maxSlugLength]
		slug = strings.TrimRight(slug, "_")
	}

	return slug
}

// GenerateUniqueSlug generates slug and handles collisions
func GenerateUniqueSlug(branch string, existing map[string]string) string {
	base := GenerateSlug(branch)

	// Check if slug exists for same branch (reuse it)
	if existingBranch, ok := existing[base]; ok && existingBranch == branch {
		return base
	}

	// Check collision with different branch
	slug := base
	suffix := 2
	for {
		if existingBranch, ok := existing[slug]; !ok || existingBranch == branch {
			break
		}
		slug = fmt.Sprintf("%s_%d", base, suffix)
		suffix++
	}

	return slug
}

File: tui/src/utils/slug.ts
/**
 * Branch-name-to-filesystem-slug conversion utility.
 *
 * Ports the Go `internal/worktree/slug.go` logic. Converts git branch
 * names like "feature/auth-flow" into filesystem-safe slugs like
 * "feature_auth-flow", handling collisions with numeric suffixes.
 */

/** Characters considered unsafe for filesystem slugs. */
const SLUG_UNSAFE_REGEX = /[^a-zA-Z0-9_-]+/g

/** Collapse repeated underscores into one. */
const MULTI_UNDERSCORE_REGEX = /_+/g

/** Clean up underscore-dash adjacency (e.g. "a_-b" or "a-_b"). */
const UNDERSCORE_DASH_REGEX = /_-|-_/g

/** Maximum slug length before truncation. */
const MAX_SLUG_LENGTH = 80

/**
 * Generate a filesystem-safe slug from a git branch name.
 *
 * Transformation steps:
 * 1. Replace `/` with `_`
 * 2. Remove unsafe characters
 * 3. Collapse multiple underscores
 * 4. Clean underscore-dash adjacencies
 * 5. Trim leading/trailing underscores
 * 6. Truncate to 80 characters
 *
 * @example
 * generateSlug("feature/auth-flow")  // "feature_auth-flow"
 * generateSlug("fix/bug///extra")    // "fix_bug_extra"
 */
export function generateSlug(branch: string): string {
  let slug = branch.replaceAll("/", "_")
  slug = slug.replace(SLUG_UNSAFE_REGEX, "_")
  slug = slug.replace(MULTI_UNDERSCORE_REGEX, "_")
  slug = slug.replace(UNDERSCORE_DASH_REGEX, "-")
  slug = slug.replace(/^_+|_+$/g, "")

  if (slug.length > MAX_SLUG_LENGTH) {
    slug = slug.slice(0, MAX_SLUG_LENGTH).replace(/_+$/, "")
  }

  return slug
}

/**
 * Generate a unique slug, handling collisions with existing slugs.
 *
 * If the slug already exists for the same branch, it is reused.
 * If it collides with a different branch, a numeric suffix (_2, _3, ...)
 * is appended until a unique slug is found.
 *
 * @param branch   - The branch name to slugify.
 * @param existing - Map of existing slug -> branch name.
 */
export function generateUniqueSlug(branch: string, existing: Map<string, string>): string {
  const base = generateSlug(branch)

  // Reuse slug if it already maps to the same branch
  const existingBranch = existing.get(base)
  if (existingBranch === branch) {
    return base
  }

  // Find a collision-free slug
  let slug = base
  let suffix = 2
  while (true) {
    const mapped = existing.get(slug)
    if (mapped === undefined || mapped === branch) {
      break
    }
    slug = `${base}_${suffix}`
    suffix++
  }

  return slug
}
